File: 0141-linked-list-cycle/0141-linked-list-cycle.java
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public boolean hasCycle(ListNode head) {
        ListNode slowPointer = head; 
        ListNode fastPointer = head; 
        
        while (fastPointer != null && fastPointer.next != null) {
            slowPointer = slowPointer.next;
            fastPointer = fastPointer.next.next;
            
            // if linked list has a cycle in it, slowPointer and fastPointer will meet
            if (slowPointer == fastPointer) {
                return true; 
            }
        }
        return false; 
        
    }
}

File: 0141-linked-list-cycle/0141-linked-list-cycle.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def hasCycle(self, head: Optional[ListNode]) -> bool:
        # 2 pointer will meet again if it has cycle
        slow, fast = head, head

        while fast and fast.next:
            slow = slow.next # +1 step
            fast = fast.next.next # +2 step
            if slow == fast:
                return True

        return False